source/common/network/utility.cc

```diff
@@ -547,6 +566,44 @@ Api::IoCallUint64Result Utility::readFromSocket(IoHandle& handle,
                                                 UdpPacketProcessor& udp_packet_processor,
                                                 MonotonicTime receive_time,
                                                 uint32_t* packets_dropped) {
+
+  if (handle.supportsUdpGro()) {
+    Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+    IoHandle::RecvMsgOutput output(1, packets_dropped);
+
+    // TODO(yugant): Avoid allocating 24k for each read by getting memory from UdpPacketProcessor
+    const uint64_t max_packet_size_with_gro = 16 * udp_packet_processor.maxPacketSize();
+
+    Api::IoCallUint64Result result =
+        receiveMessage(max_packet_size_with_gro, buffer, output, handle, local_address);
+
+    if (!result.ok()) {
+      return result;
+    }
+
+    const uint64_t gso_size = output.msg_[0].gso_size_;
+    ENVOY_LOG_MISC(trace, "recvmsg bytes {} with gso_size as {}", result.rc_, gso_size);
+
+    // Skip gso segmentation and proceed as a single payload.
+    if (gso_size == 0u) {
+      passPayloadToProcessor(result.rc_, std::move(buffer), std::move(output.msg_[0].peer_address_),
+                             std::move(output.msg_[0].local_address_), udp_packet_processor,
+                             receive_time);
+      return result;
+    }
+
+    // Segment the buffer read by the recvmsg syscall into gso_sized sub buffers.
+    while (buffer->length() > 0) {
+      const uint64_t bytes_to_copy = std::min(buffer->length(), gso_size);
```

Comment:
Is it an invariant that `gso_size` can ever be zero? This would livelock potentially. If it's an invariant, I suggest making it an `ASSERT` with comment. Otherwise we might want to be more defensive.

Reply:
In cases when gso_size is 0, we treat the operation as a recvmsg syscall without any segmentation and return before entering the loop here.

There is a check for the `gso_size == 0u`, just before the loop to avoid having gso_size as zero. Also since gso_size is defined as a const it will not change afterwards!